helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -68,23 +68,40 @@ public OptimalAssignment calculate(ClusterModel clusterModel) throws HelixRebala
     List<AssignableNode> nodes = new ArrayList<>(clusterModel.getAssignableNodes().values());
     Set<String> busyInstances =
         getBusyInstances(clusterModel.getContext().getBestPossibleAssignment().values());
-    // Sort the replicas so the input is stable for the greedy algorithm.
-    // For the other algorithm implementation, this sorting could be unnecessary.
-    for (AssignableReplica replica : getOrderedAssignableReplica(clusterModel)) {
+
+    // Compute overall utilization of the cluster. Capacity dimension -> total remaining capacity
+    Map<String, Integer> overallClusterRemainingCapacityMap =
+        computeOverallClusterRemainingCapacity(nodes);
+
+    // Create a wrapper for each AssignableReplica.
+    Set<AssignableReplicaWithScore> toBeAssignedReplicas =
+        clusterModel.getAssignableReplicaMap().values().stream()
+            .flatMap(replicas -> replicas.stream())
+            .map(replica -> new AssignableReplicaWithScore(replica, clusterModel))
+            .collect(Collectors.toSet());
+
+    AssignableReplicaWithScore replica =
+        getNextAssignableReplica(toBeAssignedReplicas,  overallClusterRemainingCapacityMap);
+    while (replica != null) {
+      toBeAssignedReplicas.remove(replica);
+      AssignableReplica rawReplica = replica.getRawAssignableReplica();
```

Comment:
(This is related to my comment on function naming for `AssignableReplicaWithScore`. )
`rawReplica` here is really just `replica`, and the `replica` before should be `replicaWithScore`, right? I believe that's more clear as the names reflect their classes, and `getRawAssignableReplica()` could just become `getAssignableReplica()`.

Reply:
TFTR. Good suggestion. Will update.